controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,361 @@
+	if !ok {
+		return false
+	}
+	// Check if the ControlPlane Version matches the version in the Cluster's managed topology spec.
+	version, err := contract.ControlPlane().Version().Get(cp)
+	if err != nil {
+		log.Infof("We failed on getting control plane version %v %v", err, cp)
+		return false
+	}
+	if *version != cluster.Spec.Topology.Version {
+		log.Infof("We failed on comparing control plane version %v", cp)
+		return false
+	}
+	// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+	if cluster.Spec.Topology.ControlPlane.Replicas != nil {
+		replicas, err := contract.ControlPlane().Replicas().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane replicas %v", cp)
+			return false
+		}
+		// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+		if int32(*replicas) != *cluster.Spec.Topology.ControlPlane.Replicas {
+			log.Infof("We failed on comparing control plane replicas %v", cp)
+			return false
+		}
+	}
+	log.Infof("ControlPlane passed with %v", cp)
+	if clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		cpInfra, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane infrastructure %v,  %v", err, cp)
+			return false
+		}
+		if ok := referenceExistsWithCorrectKindAndAPIVersion(cpInfra,
+			builder.GenericInfrastructureMachineTemplateKind,
+			builder.InfrastructureGroupVersion); !ok {
+			log.Infof("Kind or version wrong from InfrastructureMachineTemplate %v", cp)
+			return false
+		}
+		if _, ok := getAndEnsureLabelsAndAnnotations(*cpInfra, cluster.Name); !ok {
+			return false
+		}
+		log.Infof("ControlPlaneInfrastructure passed with %v", cpInfra)
+	}
+	return true
+}
+
+// assertMachineDeploymentsReconcile checks if the MachineDeployments:
+// 1) Are created in the correct number.
+// 2) Have the correct labels (TopologyOwned, ClusterName, MachineDeploymentName).
+// 3) Have the correct finalizer applied.
+// 4) Have the correct replicas and version.
+// 6) Have the correct Kind/APIVersion and Labels/Annotations for BoostrapRef and InfrastructureRef templates.
+func assertMachineDeploymentsReconcile(cluster *clusterv1.Cluster) bool {
+	// List all created machine deployments to ensure the expected numbers are c*reated.
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	if err := env.List(ctx, machineDeployments); err != nil {
+		log.Infof("machine deployments not listed %v", err)
+		return false
+	}
+	clusterMDs := make([]clusterv1.MachineDeployment, 0)
+	// Run through all machine deployments and add only those with the TopologyOwnedLable and the correct
+	// ClusterLabelName to the items for further testing.
+	for _, m := range machineDeployments.Items {
+		// If the machineDeployment doesn't have the ClusterTopologyOwnedLabel and the ClusterLabelName ignore.
+		md := m.DeepCopy()
```

Comment:
Q: why are we deep copying?
(same below)

Reply:
This is because of an 'implicit memory aliasing in for loop' issue thrown up by the linter - I dont' think it matters for this code but it could in other contexts. DeepCopy is the easiest way out,  but I'd be happy to do something different

Reply:
Usually for this case we just use something like `md := m`. Maybe that's better to avoid the impression that we need a full deep copy.